wpimath/src/main/java/edu/wpi/first/math/trajectory/ExponentialProfile.java

```diff
@@ -0,0 +1,452 @@
+   * @param goal The desired state when the profile is complete.
+   * @return The input applied at time t.
+   */
+  public double calculateInput(double t, State current, State goal) {
+    var direction = shouldFlipAcceleration(current, goal) ? -1 : 1;
+    var U = direction * m_constraints.maxInput;
+
+    var inflectionPoint = calculateInflectionPoint(current, goal, U);
+    var timing = calculateProfileTiming(current, inflectionPoint, goal, U);
+
+    if (t < 0) {
+      return 0;
+    } else if (t < timing.inflectionTime) {
+      return direction * m_constraints.maxInput;
+    } else if (t < timing.totalTime) {
+      return -direction * m_constraints.maxInput;
+    } else {
+      return 0;
+    }
+  }
+
+  /**
+   * Calculate the correct position and velocity for the profile at a time t where the current state
+   * is at time t = 0.
+   *
+   * @param t The time since the beginning of the profile.
+   * @param current The current state.
+   * @param goal The desired state when the profile is complete.
+   * @return The position and velocity of the profile at time t.
+   */
+  public State calculate(double t, State current, State goal) {
+    var direction = shouldFlipAcceleration(current, goal) ? -1 : 1;
+    var U = direction * m_constraints.maxInput;
+
+    var inflectionPoint = calculateInflectionPoint(current, goal, U);
+    var timing = calculateProfileTiming(current, inflectionPoint, goal, U);
+
+    if (t < 0) {
+      return current;
+    } else if (t < timing.inflectionTime) {
+      return new State(
+          computeDistanceFromTime(t, U, current), computeVelocityFromTime(t, U, current));
+    } else if (t < timing.totalTime) {
+      return new State(
+          computeDistanceFromTime(t - timing.totalTime, -U, goal),
+          computeVelocityFromTime(t - timing.totalTime, -U, goal));
+    } else {
+      return goal;
+    }
+  }
+
+  /**
+   * Calculate the point after which the fastest way to reach the goal state is to apply input in
+   * the opposite direction.
+   *
+   * @param current The current state.
+   * @param goal The desired state when the profile is complete.
+   * @return The position and velocity of the profile at the inflection point.
+   */
+  public State calculateInflectionPoint(State current, State goal) {
+    var direction = shouldFlipAcceleration(current, goal) ? -1 : 1;
+    var U = direction * m_constraints.maxInput;
+
+    return calculateInflectionPoint(current, goal, U);
+  }
+
+  /**
+   * Calculate the point after which the fastest way to reach the goal state is to apply input in
+   * the opposite direction.
+   *
+   * @param current The current state.
+   * @param goal The desired state when the profile is complete.
+   * @param input The signed input applied to this profile from the current state.
+   * @return The position and velocity of the profile at the inflection point.
+   */
+  private State calculateInflectionPoint(State current, State goal, double input) {
+    var U = input;
```

Comment:
Use uppercase for 2D matrices and lowercase for row vectors or column vectors (state and input are column vectors).

Reply:
done